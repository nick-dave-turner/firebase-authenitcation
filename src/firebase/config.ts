/* istanbul ignore file */
interface FirebaseConfig {
  apiKey: string | undefined;
  appId: string | undefined;
  authDomain: string | undefined;
  databaseURL: string | undefined;
  measurementId: string | undefined;
  messagingSenderId: string | undefined;
  projectId: string | undefined;
  storageBucket: string | undefined;
}

export const firebaseConfig: FirebaseConfig = {
  apiKey: process.env.REACT_APP_API_KEY,
  appId: process.env.REACT_APP_APP_ID,
  authDomain: process.env.REACT_APP_AUTH_DOMAIN,
  databaseURL: process.env.REACT_APP_DATABASE_URL,
  measurementId: process.env.REACT_APP_MEASUREMENT_ID,
  messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
  projectId: process.env.REACT_APP_PROJECT_ID,
  storageBucket: process.env.REACT_APP_PROJECT_ID,
};
