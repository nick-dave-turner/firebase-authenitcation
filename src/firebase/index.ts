/* istanbul ignore file */
export { AuthProvider, useAuthContext } from "./context";
export { firebaseAuth } from "./init";
export { useAuth } from "./use-auth";
