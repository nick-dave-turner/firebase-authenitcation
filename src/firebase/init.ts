/* istanbul ignore file */
import firebase from "firebase/app";
import "firebase/auth";

import { firebaseConfig } from "./config";

export const firebaseApp = firebase.initializeApp(firebaseConfig);
export const firebaseAuth = firebaseApp.auth();
