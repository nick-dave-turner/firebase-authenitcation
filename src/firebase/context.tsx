import React from "react";

import { firebaseAuth } from "./init";

type AuthType = { user: firebase.User | undefined };

const AuthContext = React.createContext<AuthType>({ user: undefined });

export const AuthProvider: React.FC = ({ children }) => {
  const [user, setUser] = React.useState<AuthType>({ user: undefined });
  const [initializing, setInitializing] = React.useState<boolean>(true);

  React.useEffect(() => {
    const subscriber = firebaseAuth.onAuthStateChanged((result) => {
      setUser({ user: result || undefined });
      if (initializing) setInitializing(false);
    });

    return subscriber;
  }, [initializing]);

  if (initializing) return <h1>Loading...</h1>;

  return <AuthContext.Provider value={user}>{children}</AuthContext.Provider>;
};

export const useAuthContext = (): AuthType => React.useContext(AuthContext);
