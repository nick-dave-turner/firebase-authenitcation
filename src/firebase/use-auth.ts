import { useHistory } from "react-router-dom";
import { firebaseAuth } from "./init";

interface Error {
  error?: string;
}

interface AuthWithEmail {
  email: string;
  password: string;
}

interface UseAuth {
  signInWithEmail: ({ email, password }: AuthWithEmail) => Promise<Error>;
  signUpWithEmail: ({ email, password }: AuthWithEmail) => Promise<Error>;
  signOut: () => Promise<void>;
}

export const FAILED_MESSAGE =
  "Sorry, there was an error verifying your account details. Please try again.";

export const useAuth = (): UseAuth => {
  const history = useHistory();

  const signInWithEmail = ({
    email,
    password,
  }: AuthWithEmail): Promise<Error> => {
    return new Promise((resolve, reject) => {
      firebaseAuth
        .signInWithEmailAndPassword(email, password)
        .catch(() => reject(FAILED_MESSAGE));
    });
  };

  const signUpWithEmail = ({
    email,
    password,
  }: AuthWithEmail): Promise<{ error?: string }> => {
    return new Promise((resolve, reject) => {
      firebaseAuth
        .createUserWithEmailAndPassword(email, password)
        .catch(({ message }) => reject(message));
    });
  };

  const signOut = (): Promise<void> => {
    return firebaseAuth
      .signOut()
      .then(() => history.push("/signin"))
      .catch((error) => error);
  };

  return {
    signInWithEmail,
    signUpWithEmail,
    signOut,
  };
};
