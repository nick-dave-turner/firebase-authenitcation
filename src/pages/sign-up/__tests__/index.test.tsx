import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import pretty from "pretty";

import { useAuth, useAuthContext } from "../../../firebase";
import { SignUpPage } from "..";

jest.mock("../../../firebase");

const MOCK_SIGNUP_WITH_EMAIL = jest.fn();

(useAuth as jest.Mock).mockImplementation(() => ({
  signUpWithEmail: MOCK_SIGNUP_WITH_EMAIL,
}));

describe("SignUpPage", () => {
  describe("When authenticated", () => {
    let MOCK_LOCATION: { pathname: string } = { pathname: "" };

    beforeEach(() => {
      (useAuthContext as jest.Mock).mockReturnValue({
        user: { authenticated: true },
      });
    });

    it("should redirect to home page", () => {
      render(
        <MemoryRouter initialEntries={["/signup"]}>
          <SignUpPage />
          <Route
            path="*"
            render={({ location }) => {
              MOCK_LOCATION = location;
              return "";
            }}
          />
        </MemoryRouter>
      );

      expect(MOCK_LOCATION.pathname).toBe("/home");
    });
  });

  describe("When not authenticated", () => {
    beforeEach(() => {
      (useAuthContext as jest.Mock).mockReturnValue({ user: undefined });
    });

    it("should match snapshot", () => {
      const { container } = render(
        <MemoryRouter initialEntries={["/signup"]}>
          <SignUpPage />
        </MemoryRouter>
      );

      expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
  });
});
