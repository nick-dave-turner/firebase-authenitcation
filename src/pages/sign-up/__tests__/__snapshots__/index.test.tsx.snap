// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPage When not authenticated should match snapshot 1`] = `
"<main class=\\"MuiContainer-root makeStyles-layout-1 MuiContainer-maxWidthSm\\">
  <div class=\\"MuiPaper-root makeStyles-paper-2 MuiPaper-elevation3 MuiPaper-rounded\\">
    <h1 class=\\"MuiTypography-root MuiTypography-h5\\">Welcome to Firebase Demo</h1>
    <h2 class=\\"MuiTypography-root MuiTypography-subtitle1\\">Create an account to start using Firebase Demo</h2>
    <form class=\\"makeStyles-form-3\\" novalidate=\\"\\">
      <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"email\\" id=\\"email-label\\">Email Address<span aria-hidden=\\"true\\" class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
        <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" autocomplete=\\"email\\" id=\\"email\\" name=\\"email\\" required=\\"\\" type=\\"text\\" data-testid=\\"email-input\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div>
      </div>
      <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"password\\" id=\\"password-label\\">Password<span aria-hidden=\\"true\\" class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
        <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" autocomplete=\\"current-password\\" id=\\"password\\" name=\\"password\\" required=\\"\\" type=\\"password\\" data-testid=\\"password-input\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div>
      </div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" data-testid=\\"submit-button\\" disabled=\\"\\"><span class=\\"MuiButton-label\\">Sign Up</span></button>
    </form>
    <div class=\\"MuiGrid-root MuiGrid-container\\">
      <div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true\\"></div>
      <div class=\\"MuiGrid-root MuiGrid-item\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary\\" href=\\"/signin\\">Already have an account? Sign In</a></div>
    </div>
  </div>
</main>"
`;
